### PR TITLE
Rjm/pretty print

### PR DESCRIPTION
**Pull request by rjmac**

### CODE REVIEW

File: soql-standalone-parser/src/main/scala/com/socrata/soql/ast/Select.scala [hunk 1]
```diff
@@ -357,6 +390,23 @@ case class Selection(allSystemExcept: Option[StarSelection], allUserExcept: Seq[
...
+  def doc = docs.punctuate(d",").hsep
+
+  def docs: Seq[Doc[Nothing]] = {
+    val docs = Seq.newBuilder[Doc[Nothing]]
+    docs ++= allSystemExcept.map(_.doc(":*"))
+    docs ++= allUserExcept.map(_.doc("*"))
+    docs ++= expressions.map { case SelectedExpression(expr, name) =>
+      name match {
+        case Some((n, _)) =>
+          d"${expr.doc} AS `${n.toString}`"
```
Comment (sjzavaleta) [inline soql-standalone-parser/src/main/scala/com/socrata/soql/ast/Select.scala:403]: How come this string gets formatted normally, not through docs?

File: soql-standalone-parser/src/test/scala/com/socrata/soql/parsing/ParserTest.scala [hunk 2]
```diff
@@ -231,32 +231,32 @@ class ParserTest extends WordSpec with MustMatchers {
 
     "count(*) window" in {
       val x = parseFull("select count(*) over (partition by column)")
-      x.toString must be("SELECT count(*) OVER ( PARTITION BY `column`)")
+      x.toString must be("SELECT count(*) OVER (PARTITION BY `column`)")
     }
 
     "count(column) window" in {
       val x = parseFull("select count(column_a) over (partition by column_b)")
-      x.toString must be("SELECT count(`column_a`) OVER ( PARTITION BY `column_b`)")
```
Comment (sjzavaleta) [inline soql-standalone-parser/src/test/scala/com/socrata/soql/parsing/ParserTest.scala:239]: interesting improvement!

File: soql-standalone-parser/src/main/scala/com/socrata/soql/ast/Expression.scala [hunk 3]
```diff
@@ -171,200 +161,147 @@ object SpecialFunctions {
...
-        parameters(0).format(d, sb, limit).map(_.append(" IS NULL"))
+        d"${parameters(0).doc} IS NULL"
       case SpecialFunctions.IsNotNull =>
-        parameters(0).format(d, sb, limit).map(_.append(" IS NOT NULL"))
+        d"${parameters(0).doc} IS NOT NULL"
       case SpecialFunctions.In =>
-        parameters(0).format(d, sb, limit).map(_.append(" IN (")).flatMap { sb =>
-          appendParams(sb, parameters.iterator.drop(1), ",", d, limit)
-        }.map(_.append(")"))
-      case SpecialFunctions.NotIn =>
```
Comment (sjzavaleta) [inline soql-standalone-parser/src/main/scala/com/socrata/soql/ast/Expression.scala:305]: is NotIN not an available function? Maybe it gets interpreted as a NOT and then an IN?

File: soql-standalone-parser/src/main/scala/com/socrata/soql/ast/Expression.scala [hunk 4]
```diff
@@ -399,8 +349,18 @@ case class WindowFunctionInfo(partitions: Seq[Expression], orderings: Seq[OrderB
...
   def allColumnRefs = Set.empty
-  protected override def asString = "?" + name
-  def format(depth: Int, sb: StringBuilder, limit: Option[Int]) = Some(sb.append(asString))
 
   def replaceHoles(f: Hole => Expression) = f(this)
+
+  def doc = Doc("?" + name)
+}
+
+object bleh extends App {
```
Comment (sjzavaleta) [inline soql-standalone-parser/src/main/scala/com/socrata/soql/ast/Expression.scala:358]: Was this for testing?

File: soql-standalone-parser/src/main/scala/com/socrata/soql/ast/Expression.scala (hunk 3 above)
Comment (rjmac) [reply soql-standalone-parser/src/main/scala/com/socrata/soql/ast/Expression.scala:305]: Er, nope, I just missed it.

File: soql-standalone-parser/src/main/scala/com/socrata/soql/ast/Select.scala (hunk 1 above)
Comment (rjmac) [reply soql-standalone-parser/src/main/scala/com/socrata/soql/ast/Select.scala:403]: The `d` interpolator can do that for you.  I was experimenting with that and then decided I preferred `d` as _just_ a string literal syntax, but it actually is a full interpolator.  This is precisely equivalent to ```expr.doc ++ d" AS `" ++ Doc(n.toString) ++ d"`"```.  I'll probably go back and change it to the other style.

File: soql-standalone-parser/src/main/scala/com/socrata/soql/ast/Expression.scala (hunk 4 above)
Comment (rjmac) [reply soql-standalone-parser/src/main/scala/com/socrata/soql/ast/Expression.scala:358]: Yeah, oops.  Well, kinda.  This is the App that powers that demo site I put up to show the formatter off. :)

**Review by sjzavaleta — APPROVED** [review]

**Pull request merged**